src/history/HistoryArchive.cpp

```diff
@@ -275,17 +276,35 @@ HistoryArchiveState::HistoryArchiveState() : server(STELLAR_CORE_VERSION)
 }
 
 HistoryArchiveState::HistoryArchiveState(uint32_t ledgerSeq,
-                                         BucketList const& buckets)
+                                         BucketList const& buckets,
+                                         bool clearFutureBuckets)
     : server(STELLAR_CORE_VERSION), currentLedger(ledgerSeq)
 {
+    std::shared_ptr<Bucket> prevSnap;
     for (uint32_t i = 0; i < BucketList::kNumLevels; ++i)
     {
         HistoryStateBucket b;
         auto& level = buckets.getLevel(i);
         b.curr = binToHex(level.getCurr()->getHash());
-        b.next = level.getNext();
         b.snap = binToHex(level.getSnap()->getHash());
+        b.next = level.getNext();
+
+        // Level 0 merge always happens in-place on close ledger, so next always
+        // remains clear in between `addBatch` calls
+        if (clearFutureBuckets && prevSnap)
+        {
+            BucketInputIterator it(prevSnap);
```

Comment:
You're doing this "make an iterator, sniff its version" thing in a couple places, it might be worth factoring out into a helper method on a bucket.

(Part of me also thinks maybe it should be cached, but buckets are accessed from multiple threads, and .. meh, it's probably not especially hot)

Reply:
fixed